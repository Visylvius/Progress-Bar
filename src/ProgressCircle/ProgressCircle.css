.ProgressCircle-background,
.ProgressCircle-progressPercentage {
  fill: none;
}

.ProgressCircle-background {
  stroke: #ddd;
}

.ProgressCircle-progressPercentage {
  stroke: #439df4;
  stroke-linecap: round;
  stroke-linejoin: round;
}

.ProgressCircle-percentageText {
  font-size: 4em;
  fill: #303030;
}

.ProgressCircle-percentageSign {
  fill: #ddd;
  font-size: 1.5em;
}
