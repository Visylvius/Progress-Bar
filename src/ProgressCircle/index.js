import ProgressCirle from "./ProgressCircle";

export default ProgressCirle;
