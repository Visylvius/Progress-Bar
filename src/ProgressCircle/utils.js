/**
 *
 * @param {number} percentage - current percentage of the widget
 * @returns {string}
 */
function getXAttribute(percentage) {
  if (percentage < 10) {
    return "60%";
  } else if (percentage <= 99) {
    return "70%";
  }
  return "77%";
}

export { getXAttribute };
