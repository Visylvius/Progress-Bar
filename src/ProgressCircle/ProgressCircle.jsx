import React from "react";

import "./ProgressCircle.css";

/** ProgressCircle Props
 *
 * @param {number} sqSize Size of the enclosing square
 * @param {number} strokeWidth Width of the SVG stroke
 * @param {number} percentage number responsible for showing where the indicator is currently at
 */

const ProgressCircle = ({ sqSize = 200, strokeWidth = 15, percentage = 0 }) => {
  // SVG centers the stroke width on the radius, subtract out so circle fits in square
  const radius = (sqSize - strokeWidth) / 2;
  // Enclose circle in a circumscribing square
  const viewBox = `0 0 ${sqSize} ${sqSize}`;
  // Arc length at 100% coverage is the circle circumference
  const dashArray = radius * Math.PI * 2;
  // Scale 100% coverage overlay with the actual percent
  const dashOffset = dashArray - (dashArray * percentage) / 100;

  return (
    <svg width={sqSize} height={sqSize} viewBox={viewBox}>
      <circle
        className="ProgressCircle-background"
        cx={sqSize / 2}
        cy={sqSize / 2}
        r={radius}
        strokeWidth={`${strokeWidth}px`}
        data-testid="ProgressCircle-background"
      />
      <circle
        className="ProgressCircle-progressPercentage"
        cx={sqSize / 2}
        cy={sqSize / 2}
        r={radius}
        strokeWidth={`${strokeWidth}px`}
        // Start progress marker at 12 O'Clock
        transform={`rotate(-90 ${sqSize / 2} ${sqSize / 2})`}
        style={{
          strokeDasharray: dashArray,
          strokeDashoffset: dashOffset,
        }}
        data-testid="ProgressCircle-progressPercentage"
      />
      <text
        className="ProgressCircle-percentageText"
        x="45%"
        y="50%"
        dy=".3em"
        textAnchor="middle"
      >
        {`${percentage}`}
      </text>
      <text
        className="ProgressCircle-percentageSign"
        x={percentage >= 10 ? "68%" : "60%"}
        y="40%"
        dy=".3em"
        textAnchor="middle"
      >
        %
      </text>
    </svg>
  );
};

export default ProgressCircle;
