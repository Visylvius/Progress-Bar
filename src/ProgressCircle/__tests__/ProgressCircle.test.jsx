import React from "react";
import { render, screen } from "@testing-library/react";

import ProgressCircle from "../ProgressCircle";

describe("ProgressCircle (Unit)", () => {
  const renderProgressCircle = (props) => render(<ProgressCircle {...props} />);

  describe("render", () => {
    test("should match snapshot", () => {
      const { asFragment } = renderProgressCircle();

      expect(asFragment()).toMatchSnapshot();
    });

    describe("ProgressCircle Background Element", () => {
      test("should have the proper cx, cy, and radius values", () => {
        renderProgressCircle({ sqSize: 100, strokeWidth: 15 });
        const progressCircle = screen.getByTestId("ProgressCircle-background");

        expect(progressCircle.getAttribute("cx")).toBe("50");
        expect(progressCircle.getAttribute("cy")).toBe("50");
        expect(progressCircle.getAttribute("r")).toBe("42.5");
      });
    });

    describe("ProgressCircle Percentage Bar", () => {
      test("should have the proper cx, cy, and radius values", () => {
        renderProgressCircle({ sqSize: 50, strokeWidth: 15 });
        const percentageCircle = screen.getByTestId(
          "ProgressCircle-progressPercentage"
        );

        expect(percentageCircle.getAttribute("cx")).toBe("25");
        expect(percentageCircle.getAttribute("cy")).toBe("25");
        expect(percentageCircle.getAttribute("r")).toBe("17.5");
      });

      test("should have the proper transform value", () => {
        renderProgressCircle({ sqSize: 50, strokeWidth: 15 });
        const percentageCircle = screen.getByTestId(
          "ProgressCircle-progressPercentage"
        );

        expect(percentageCircle.getAttribute("transform")).toBe(
          "rotate(-90 25 25)"
        );
      });

      test("should have the proper style tags", () => {
        renderProgressCircle({ sqSize: 50, strokeWidth: 15 });
        const percentageCircle = screen.getByTestId(
          "ProgressCircle-progressPercentage"
        );

        expect(percentageCircle.getAttribute("style")).toBe(
          "stroke-dasharray: 109.95574287564276; stroke-dashoffset: 109.95574287564276;"
        );
      });
    });
  });
});
