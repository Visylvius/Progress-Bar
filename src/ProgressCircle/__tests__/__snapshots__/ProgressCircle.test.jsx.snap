// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressCircle (Unit) render should match snapshot 1`] = `
<DocumentFragment>
  <svg
    height="200"
    viewBox="0 0 200 200"
    width="200"
  >
    <circle
      class="ProgressCircle-background"
      cx="100"
      cy="100"
      data-testid="ProgressCircle-background"
      r="92.5"
      stroke-width="15px"
    />
    <circle
      class="ProgressCircle-progressPercentage"
      cx="100"
      cy="100"
      data-testid="ProgressCircle-progressPercentage"
      r="92.5"
      stroke-width="15px"
      style="stroke-dasharray: 581.1946409141117; stroke-dashoffset: 581.1946409141117;"
      transform="rotate(-90 100 100)"
    />
    <text
      class="ProgressCircle-percentageText"
      dy=".3em"
      text-anchor="middle"
      x="45%"
      y="50%"
    >
      0
    </text>
    <text
      class="ProgressCircle-percentageSign"
      dy=".3em"
      text-anchor="middle"
      x="60%"
      y="40%"
    >
      %
    </text>
  </svg>
</DocumentFragment>
`;
